src/config.rs

```diff
@@ -33,44 +34,123 @@ impl Shell {
-    #[serde(default = "Vec::new")]
     pub(crate) init: Vec<String>,
-    #[serde(default = "default_shell")]
     pub(crate) shell: Shell,
-    #[serde(default = "default_mount")]
-    pub(crate) mount: String,
-    #[serde(default = "Vec::new")]
+
+    pub(crate) floki_working_directory: path::PathBuf,
+    pub(crate) host_mount_directory: path::PathBuf,
+    pub(crate) guest_mount_directory: path::PathBuf,
+    pub(crate) guest_working_directory: path::PathBuf,
+
     pub(crate) docker_switches: Vec<String>,
-    #[serde(default = "default_to_false")]
     pub(crate) forward_ssh_agent: bool,
-    #[serde(default = "default_to_false")]
     pub(crate) dind: bool,
-    #[serde(default = "default_to_false")]
     pub(crate) forward_user: bool,
 }
 
 impl FlokiConfig {
-    pub fn from_file(file: &str) -> Result<FlokiConfig, Error> {
+    pub fn from_file(
+        floki_working_directory: path::PathBuf,
+        config_file: Option<path::PathBuf>,
+    ) -> Result<FlokiConfig, Error> {
+        let (host_mount_directory, config_file) = match config_file {
+            Some(file) => (floki_working_directory.to_path_buf(), file),
+            None => {
+                let config_file = find_file_in_tree(&floki_working_directory, "floki.yaml")?;
+                // config_file is guaranteed to be a file not a directory
+                (config_file.parent().unwrap().to_path_buf(), config_file)
+            }
+        };
+        debug!("Selected configuration file: {:?}", &config_file);
+
+        let static_config = StaticFlokiConfig::from_file(&config_file)?;
+        debug!("Got configuration: {:?}", &static_config);
+
+        let guest_working_directory = path::PathBuf::from(&static_config.mount).join(
+            // The host mount directory is guaranteed to be an ancestor of the
+            // floki working directory.
+            floki_working_directory
+                .strip_prefix(&host_mount_directory)
+                .unwrap(),
+        );
+        debug!(
+            "Using guest working directory: {:?}",
+            &guest_working_directory
+        );
+
+        Ok(FlokiConfig {
+            image: static_config.image,
+            init: static_config.init,
+            shell: static_config.shell,
+            floki_working_directory,
+            host_mount_directory,
+            guest_mount_directory: static_config.mount.into(),
+            guest_working_directory,
+            docker_switches: static_config.docker_switches,
+            forward_ssh_agent: static_config.forward_ssh_agent,
+            dind: static_config.dind,
+            forward_user: static_config.forward_user,
+        })
+    }
+}
+
+/// Search all ancestor directories of start_filepath for a file of the given
+/// name.
+fn find_file_in_tree(start_filepath: &path::Path, filename: &str) -> Result<path::PathBuf, Error> {
+    start_filepath
+        .ancestors()
+        .map(|a| a.join(filename))
+        .filter(|f| f.is_file())
```

Comment:
I think `find` might be the iterator function you want here (or something like that)

Reply:
Ah, yep, that's perfect!